### PR TITLE
Add container_resolution

### PR DESCRIPTION
**Pull request by cat-bro**
Add container_resolution index, show, resolve, resolve_toolbox, resolve_toolbox_with_install.

I'm making this a draft because it still needs examples and tests.

### CODE REVIEW

**Comment by bernt-matthias** [comment]
Cool stuff. Certainly needed.

File: bioblend/galaxy/container_resolution/__init__.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        """
+        params = {}
+        if tool_ids:
+            params["tool_ids"] = ",".join(tool_ids)
+        if resolver_type:
+            params["resolver_type"] = resolver_type
+        if container_type:
+            params["container_type"] = container_type
+        params["requirements_only"] = requirements_only
+        params["install"] = install
```
Comment (bernt-matthias) [inline bioblend/galaxy/container_resolution/__init__.py:118]:
```suggestion
        params["requirements_only"] = str(requirements_only)
        params["install"] = str(install)
```

File: bioblend/galaxy/container_resolution/__init__.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
```
Comment (bernt-matthias) [inline bioblend/galaxy/container_resolution/__init__.py:190]:
```suggestion
        params["requirements_only"] = str(requirements_only)
        params["install"] = str(install)
```

**Review by bernt-matthias — COMMENTED** [review]
That's done in `ToolDependenciesClient`

**Review by bernt-matthias — COMMENTED** [review]
Did some first experiments with `resolve_toolbox` using default parameters .. looks very good :)

File: bioblend/galaxy/container_resolution/__init__.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        return self._get(url=url, params=params)
+
+    def resolve_toolbox_with_install(
+        self,
+        index: Optional[int] = None,
+        tool_ids: Optional[List[str]] = None,
+        resolver_type: Optional[str] = None,
+        container_type: Optional[str] = None,
+        requirements_only: bool = False,
+        install: bool = False,
```
Comment (bernt-matthias) [inline bioblend/galaxy/container_resolution/__init__.py:132]:
```suggestion
```

File: bioblend/galaxy/container_resolution/__init__.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        :param resolver_type: restrict to specified resolver type
+
+        :type container_type: str
+        :param container_type: restrict to specified container type
+
+        :type requirements_only: bool
+        :param requirements_only: ignore tool containers, properties - just search based on tool requirements set to True to mimic default behavior of tool dependency API.
+
+        :type install: bool
+        :param install: allow installation of new containers (for build_mulled* containers) the way job resolution will operate, defaults to False
```
Comment (bernt-matthias) [inline bioblend/galaxy/container_resolution/__init__.py:154]:
```suggestion
```

File: bioblend/galaxy/container_resolution/__init__.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        return self._get(url=url, params=params)
+
+    def resolve_toolbox(
+        self,
+        index: Optional[int] = None,
+        tool_ids: Optional[List[str]] = None,
+        resolver_type: Optional[str] = None,
+        container_type: Optional[str] = None,
+        requirements_only: bool = False,
+        install: bool = False,
```
[the remark this answers is not in the document]
Comment (cat-bro) [reply bioblend/galaxy/container_resolution/__init__.py:81]: That confuses me greatly. For the get requests I feel that they can't possibly mean it.

[the remark this answers is not in the document]
Comment (bernt-matthias) [reply bioblend/galaxy/container_resolution/__init__.py:81]: According to the API docs the only argument is `tool_ids` https://github.com/galaxyproject/galaxy/blob/6f50816fa5d91fa744b1f44df169dad651b4e10a/lib/galaxy/webapps/galaxy/api/container_resolution.py#L81. 

This seems to be in accordance with the called merthod `ContainerResolutionView.resolve_toolbox` which also only uses `tool_ids` and passes it to tre `resolve` method

**Comment by bernt-matthias** [comment]
Hey @cat-bro and @nsoranzo .. I would like to finish this. 

Unfortunately tests fail for `< 22.05`, e.g. I find the following in the logs:

```
Traceback (most recent call last):
  File "/home/berntm/projects/galaxy/lib/galaxy/web/framework/decorators.py", line 320, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "/home/berntm/projects/galaxy/lib/galaxy/web/framework/decorators.py", line 112, in decorator
    return func(self, trans, *args, **kwargs)
  File "/home/berntm/projects/galaxy/lib/galaxy/webapps/galaxy/api/container_resolution.py", line 31, in index
    return self._view.index()
  File "/home/berntm/projects/galaxy/lib/galaxy/tool_util/deps/views.py", line 328, in index
    return [r.to_dict() for r in self._container_resolvers]
  File "/home/berntm/projects/galaxy/lib/galaxy/tool_util/deps/views.py", line 399, in _container_resolvers
    return self._app.container_finder.container_resolvers
AttributeError: 'ContainerFinder' object has no attribute 'container_resolvers'
```

Seems that this did not work before 22.05: https://github.com/galaxyproject/galaxy/pull/14859

File: bioblend/galaxy/container_resolution/__init__.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        :type install: bool
+        :param install: allow installation of new containers (for build_mulled containers) the way job resolution will operate, defaults to False
+
+        :rtype: list
+          For example::
+        [{'tool_id': 'upload1', 'status': {'model_class': 'NullDependency', 'dependency_type': None, 'exact': True, 'name': None, 'version': None, 'cacheable': False}, 'requirements': []}, ...]
+        """
+        params = {}
+        if tool_ids:
+            params["tool_ids"] = ",".join(tool_ids)
```
Comment (mvdbeek) [inline bioblend/galaxy/container_resolution/__init__.py:159]: You can just pass in a list, but I'd send it as json.

File: bioblend/galaxy/container_resolution/__init__.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+            'environment_path': 'docker://quay.io/biocontainers/canu:2.2--ha47f30e_0',
+            'exact': True,
+            'model_class': 'ContainerDependency',
+            'name': None,
+            'version': None},
+        'tool_id': 'toolshed.g2.bx.psu.edu/repos/bgruening/canu/canu/2.2+galaxy0'}]
+        """
+        params = {}
+        if tool_ids:
+            params["tool_ids"] = ",".join(tool_ids)
```
Comment (mvdbeek) [inline bioblend/galaxy/container_resolution/__init__.py:228]: You can just pass in a list, but I'd send it as json.

**Comment by bernt-matthias** [comment]
Huh, did you merge by accident?

File: bioblend/galaxy/container_resolution/__init__.py (hunk 6 above)
Comment (bernt-matthias) [reply bioblend/galaxy/container_resolution/__init__.py:159]: Is there somewhere an example, of how to do this? That would be cool .. in my tests I already had problems with the length.

Comment (mvdbeek) [reply bioblend/galaxy/container_resolution/__init__.py:159]: Hmm, there's some wrapping going on here around the requests library. You'd normally set `json=True` and there's the caveat of https://stackoverflow.com/questions/978061/http-get-with-request-body. Given the things this call does maybe a POST would have been more appropriate.

**Pull request merged**